api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java

```diff
@@ -0,0 +1,324 @@
+  private volatile ScheduledFuture<?> tokenRefreshFuture;
+
+  OAuth2Client(OAuth2ClientParams params) {
+    grantType = params.getGrantType();
+    username = params.getUsername().orElse(null);
+    password = params.getPassword().orElse(null);
+    scope = params.getScope().orElse(null);
+    defaultAccessTokenLifespan = params.getDefaultAccessTokenLifespan();
+    defaultRefreshTokenLifespan = params.getDefaultRefreshTokenLifespan();
+    refreshSafetyWindow = params.getRefreshSafetyWindow();
+    tokenExchangeEnabled = params.getTokenExchangeEnabled();
+    httpClient = params.getHttpClient().addResponseFilter(this::checkErrorResponse).build();
+    executor = params.getExecutor().orElseGet(OAuth2Client::createDefaultExecutor);
+    shouldCloseExecutor = !params.getExecutor().isPresent();
+    objectMapper = params.getObjectMapper();
+    currentTokensStage = started.thenApplyAsync((v) -> fetchNewTokens(), executor);
+    currentTokensStage.thenAccept(this::scheduleTokensRenewal);
+  }
+
+  private static ScheduledExecutorService createDefaultExecutor() {
+    return Executors.newSingleThreadScheduledExecutor(
+        runnable -> {
+          Thread thread = Executors.defaultThreadFactory().newThread(runnable);
+          thread.setDaemon(true);
+          thread.setName("oauth2-client");
+          return thread;
+        });
+  }
+
+  @Override
+  public AccessToken authenticate() {
+    return getCurrentTokens().getAccessToken();
+  }
+
+  /** Visible for testing. */
+  Tokens getCurrentTokens() {
+    try {
+      return currentTokensStage.toCompletableFuture().get();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (ExecutionException e) {
+      Throwable cause = e.getCause();
+      if (cause instanceof RuntimeException) {
+        throw (RuntimeException) cause;
+      } else if (cause instanceof Error) {
+        throw (Error) cause;
+      } else {
+        throw new RuntimeException(cause);
+      }
+    }
+  }
+
+  /**
+   * Starts the client.
+   *
+   * <p>Calling this method will trigger a first access token fetch, using the "client_credentials"
+   * grant type. It will also schedule a refresh of the access token when needed.
+   *
+   * <p>Calling this method twice or more is a no-op.
+   */
+  public void start() {
+    started.complete(null);
+  }
+
+  @Override
+  public void close() {
+    try {
+      currentTokensStage.toCompletableFuture().cancel(true);
+      ScheduledFuture<?> tokenRefreshFuture = this.tokenRefreshFuture;
+      if (tokenRefreshFuture != null) {
+        tokenRefreshFuture.cancel(true);
+      }
+      if (shouldCloseExecutor) {
+        if (!executor.isShutdown()) {
+          executor.shutdown();
+          if (!executor.awaitTermination(10, TimeUnit.SECONDS)) {
+            executor.shutdownNow();
+          }
+        }
+      }
+      if (password != null) {
```

Comment:
Hm - if the same password's used in multiple clients/authentictors  (probably never happens though, but tests maybe...), a `close()` would corrupt the password for other instances.

Reply:
Good catch, but that doesn't happen anymore, because in `OAuth2ClientParams` we are using plain strings now. This class creates a new array when reading the password.